peer/abstractlist/list.go

```diff
@@ -0,0 +1,592 @@
+	for _, id := range updates.Removals {
+		errs = multierr.Append(errs, pl.remove(id))
+	}
+
+	add := updates.Additions
+	if !pl.noShuffle {
+		add = shuffle(pl.randSrc, add)
+	}
+
+	for _, id := range add {
+		errs = multierr.Append(errs, pl.add(id))
+	}
+	return errs
+}
+
+func (pl *List) updateOffline(updates peer.ListUpdates) error {
+	var errs error
+	for _, id := range updates.Removals {
+		errs = multierr.Append(errs, pl.removeOffline(id))
+	}
+	for _, id := range updates.Additions {
+		errs = multierr.Append(errs, pl.addOffline(id))
+	}
+	return errs
+}
+
+// Start notifies the List that requests will start coming
+func (pl *List) Start() error {
+	return pl.once.Start(pl.start)
+}
+
+func (pl *List) start() error {
+	pl.lock.Lock()
+	defer pl.lock.Unlock()
+
+	all := pl.offlinePeerIdentifiers()
+
+	var err error
+	err = multierr.Append(err, pl.updateOffline(peer.ListUpdates{
+		Removals: all,
+	}))
+	err = multierr.Append(err, pl.updateOnline(peer.ListUpdates{
+		Additions: all,
+	}))
+	return err
+}
+
+// Stop notifies the List that requests will stop coming
+func (pl *List) Stop() error {
+	return pl.once.Stop(pl.stop)
+}
+
+func (pl *List) stop() error {
+	pl.lock.Lock()
+	defer pl.lock.Unlock()
+
+	all := pl.onlinePeerIdentifiers()
+
+	var err error
+	err = multierr.Append(err, pl.updateOnline(peer.ListUpdates{
+		Removals: all,
+	}))
+	err = multierr.Append(err, pl.updateOffline(peer.ListUpdates{
+		Additions: all,
+	}))
+	return err
+}
+
+// IsRunning returns whether the peer list is running.
+func (pl *List) IsRunning() bool {
+	return pl.once.IsRunning()
+}
+
+// add retains a peer and sets up a thunk (a thin proxy for a peer) to receive
+// connection status notifications from the dialer and track pending request
+// counts.
+//
+// add does not add the peer to the list of peers available for choosing (the
+// Implementation).
+// The thunk is responsible for adding and removing the peer from the
+// collection of available peers based on connection status notifications.
+// Must be run inside a mutex.Lock()
+func (pl *List) add(id peer.Identifier) error {
+	addr := id.Identifier()
+
+	if _, ok := pl.peers[addr]; ok {
+		return peer.ErrPeerAddAlreadyInList(addr)
+	}
+
+	t := newPeerFacade(pl, id)
+	t.boundOnFinish = t.onFinish
+
+	p, err := pl.transport.RetainPeer(id, t)
+	if err != nil {
+		return err
+	}
+
+	t.peer = p
+	pl.peers[addr] = t
+	t.notifyStatusChanged(p)
+
+	return nil
+}
+
+func (pl *List) addOffline(id peer.Identifier) error {
+	addr := id.Identifier()
+
+	if _, ok := pl.offlinePeers[addr]; ok {
+		return peer.ErrPeerAddAlreadyInList(addr)
+	}
+
+	pl.offlinePeers[addr] = id
+	return nil
+}
+
+// remove releases and forgets a peer.
+// Must be run in a mutex.Lock()
+func (pl *List) remove(id peer.Identifier) error {
+	addr := id.Identifier()
+
+	t, ok := pl.peers[addr]
+	if !ok {
+		return peer.ErrPeerRemoveNotInList(addr)
+	}
+
+	t.remove()
+	delete(pl.peers, addr)
+
+	return pl.transport.ReleasePeer(id, t)
```

Comment:
No, but it is safe to do so. I think I would prefer to discover a deadlock in tests if a transport tried to send a notification to the subscriber in the retain/release stack, than relax the unnecessary lock, since we want that unlock to occur even through a panic.

We’ve been bitten in the past with `defer` optimization.

Reply:
>  since we want that unlock to occur even through a panic.

Do we have a spot on the outbound side which captures panics? I didn't think so.

> We’ve been bitten in the past with defer optimization.

This is true unfortunately. Would you consider the queuing all of the peers to be released, and execute the ReleasePeer in a single goroutine?

Reply:
The inbound panic handler would catch outbound panics if they’re called in that context, allowing the process to continue with corrupted state.